packages/shared/src/events.ts

```diff
@@ -0,0 +1,106 @@
+
+export interface Eventing {
+  /**
+   * Adds a handler for the given provider event type.
+   * The handlers are called in the order they have been added.
+   * @param {ProviderEvents} eventType The provider event type to listen to
+   * @param {EventHandler} handler The handler to run on occurrence of the event type
+   */
+  addHandler(eventType: ProviderEvents, handler: EventHandler): void;
+
+  /**
+   * Removes a handler for the given provider event type.
+   * @param {ProviderEvents} eventType The provider event type to remove the listener for
+   * @param {EventHandler} handler The handler to remove for the provider event type
+   */
+  removeHandler(eventType: ProviderEvents, handler: EventHandler): void;
+
+  /**
+   * Gets the current handlers for the given provider event type.
+   * @param {ProviderEvents} eventType The provider event type to get the current handlers for
+   * @returns {EventHandler[]} The handlers currently attached to the given provider event type
+   */
+  getHandlers(eventType: ProviderEvents): EventHandler[];
+}
+
+export enum ProviderEvents {
+  /**
+   * The provider is ready to evaluate flags.
+   */
+  Ready = 'PROVIDER_READY',
+
+  /**
+   * The provider is in an error state.
+   */
+  Error = 'PROVIDER_ERROR',
+
+  /**
+   * The flag configuration in the source-of-truth has changed.
+   */
+  ConfigurationChanged = 'PROVIDER_CONFIGURATION_CHANGED',
+
+  /**
+   * The provider's cached state is not longer valid and may not be up-to-date with the source of truth.
+   */
+  Stale = 'PROVIDER_STALE',
+}
+
+export class OpenFeatureEventEmitter implements ManageLogger<OpenFeatureEventEmitter> {
+  private _eventLogger?: Logger;
+  private eventEmitter = new EventEmitter({ captureRejections: true });
+
+  constructor(private readonly globalLogger?: () => Logger) {
+    this.eventEmitter.on('error', (err) => {
+      this._logger?.error('Error running event handler:', err);
+    });
+  }
+
+  emit(eventType: ProviderEvents, context?: EventDetails): void {
+    this.eventEmitter.emit(eventType, context);
+  }
+
+  addHandler(eventType: ProviderEvents, handler: EventHandler): void {
+    this.eventEmitter.on(eventType, handler);
+  }
+
+  removeHandler(eventType: ProviderEvents, handler: EventHandler): void {
+    this.eventEmitter.removeListener(eventType, handler);
+  }
+
+  removeAllHandlers(eventType?: ProviderEvents): void {
+    // If giving undefined, the listeners are not removed, so we have to check explicitly
+    if (eventType) {
+      this.eventEmitter.removeAllListeners(eventType);
+    } else {
+      this.eventEmitter.removeAllListeners();
+    }
+  }
+
+  getHandlers(eventType: ProviderEvents): EventHandler[] {
+    return this.eventEmitter.listeners(eventType) as EventHandler[];
+  }
+
+  setLogger(logger: Logger): this {
+    this._eventLogger = logger;
+    return this;
+  }
+
+  private get _logger() {
+    return this._eventLogger || this.globalLogger?.();
+  }
+}
```

Comment:
@toddbaert This would be the solution if we want sync and async handlers.
This would make it safe for javascript too.
I think I would go for this, what do you think?
```suggestion
export class OpenFeatureEventEmitter implements ManageLogger<OpenFeatureEventEmitter> {
  private readonly _handlers = new Map<EventHandler, EventHandler>();
  private readonly eventEmitter = new EventEmitter({ captureRejections: true });
  private _eventLogger?: Logger;

  constructor(private readonly globalLogger?: () => Logger) {
    this.eventEmitter.on('error', (err) => {
      this._logger?.error('Error running event handler:', err);
    });
  }

  emit(eventType: ProviderEvents, context?: EventDetails): void {
    this.eventEmitter.emit(eventType, context);
  }

  addHandler(eventType: ProviderEvents, handler: EventHandler): void {
    const asyncHandler = async (context?: EventDetails) => {
      await handler(context);
    };

    this._handlers.set(handler, asyncHandler);
    this.eventEmitter.on(eventType, asyncHandler);
  }

  removeHandler(eventType: ProviderEvents, handler: EventHandler): void {
    const asyncHandler = this._handlers.get(handler);

    if (!asyncHandler) {
      return;
    }

    this.eventEmitter.removeListener(eventType, asyncHandler);
  }

  removeAllHandlers(eventType?: ProviderEvents): void {
    // If giving undefined, the listeners are not removed, so we have to check explicitly
    if (eventType) {
      this.eventEmitter.removeAllListeners(eventType);
    } else {
      this.eventEmitter.removeAllListeners();
    }
  }

  getHandlers(eventType: ProviderEvents): EventHandler[] {
    return this.eventEmitter.listeners(eventType) as EventHandler[];
  }

  setLogger(logger: Logger): this {
    this._eventLogger = logger;
    return this;
  }

  private get _logger() {
    return this._eventLogger || this.globalLogger?.();
  }
}
```

Reply:
I think this is an elegant solution to the problem. The use of the weakmap indexed by the wrapped handler seems like a good idea.

I'm not sure the `await` on line 80 is actually needed. I don't _think_ there's a need to await handlers since we don't have any ordering guarantees. We could add a `.catch()` and log errors in the execution of the wrapped handler as well. What do you think about that?

Reply:
Nice then let us go for that solution, I will quickly add this 🙂
The wait is actually needed because we have the error listener in the constructor. We could also remove that listener and catch the error in the handler but I would personally do it as it is.

Reply:
> The wait is actually needed because we have the error listener in the constructor.

Ah yes, I see. That wasn't clear to me initially... but maybe it's better to have the `await` on line 80 anyway since we need to have the `async` keyword on the handler wrapper.

Reply:
Can you add some comments explaining why the handlers are wrapped and we're doing all this? Basically in a line or two explain what we've been talking about, just for the future.

Reply:
> but maybe it's better to have the await on line 80 anyway since we need to have the async keyword on the handler wrapper.

Does that mean we leave it as it is?

Reply:
Sure I will add a comment and add the change.

Reply:
Done, please check if the comments are okay @toddbaert 🙂

Reply:
Look great to me!